flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/CategoricalHistogram.scala

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.statistics
+
+import scala.collection.mutable
+
+/** Implementation of a discrete valued online histogram
+  *
+  * =Parameters=
+  * -[[numCategories]]:
+  *   Number of categories in the histogram
+  */
+case class CategoricalHistogram(numCategories: Int) extends OnlineHistogram {
+
+  require(numCategories > 0, "Capacity must be greater than zero")
+  val data = new mutable.HashMap[Double, Int]()
+
+  /** Number of categories in the histogram
+    *
+    * @return number of categories
+    */
+  override def bins: Int = {
+    numCategories
+  }
+
+  /** Increment count of category c
+    *
+    * @param c category whose count needs to be incremented
+    */
+  override def add(c: Double): Unit = {
+    if (data.get(c).isEmpty) {
+      require(data.size < numCategories, "Insufficient capacity. Failed to add.")
+      data.put(c, 1)
+    } else {
+      data.update(c, data.get(c).get + 1)
+    }
+  }
+
+  /** Merges the histogram with h and returns a new histogram
+    *
+    * @param h histogram to be merged
+    * @param B number of categories in the resultant histogram.
+    *          (Default: ```0```, number of categories will be the size of union of categories in
+    *          both histograms)
+    * @return Merged histogram with capacity B
+    */
+  override def merge(h: OnlineHistogram, B: Int = 0): CategoricalHistogram = {
+    h match {
+      case h1: CategoricalHistogram => {
+        val finalMap = new mutable.HashMap[Double, Int]()
+        data.iterator.foreach(x => finalMap.put(x._1, x._2))
+        h1.data.iterator.foreach(x => {
+          if (finalMap.get(x._1).isEmpty) {
+            finalMap.put(x._1, x._2)
+          } else {
+            finalMap.update(x._1, x._2 + finalMap.get(x._1).get)
+          }
+        })
```

Comment:
As I said above, we can reduce calling `get(x._1)` by using pattern matching for `finalMap.get(x._1)`.